aptos-move/framework/aptos-framework/sources/configs/Stake.move

```diff
@@ -98,58 +143,128 @@ module AptosFramework::Stake {
     }
 
     /// Unlock from active delegation, it's moved to pending_inactive if locked_until_secs < current_time or
     /// directly inactive if it's not from an active validator.
-    public(script) fun unlock(account: &signer, from: address) acquires StakePool, ValidatorSet {
-        let addr = Signer::address_of(account);
+    public(script) fun unlock(
+        delegator: &signer,
+        validator_address: address,
+    ) acquires Delegation, StakePool, ValidatorSet {
+        let delegator_address = Signer::address_of(delegator);
         let current_time = Timestamp::now_seconds();
-        let stake_pool = borrow_global_mut<StakePool>(from);
-        let delegation = withdraw_internal(&mut stake_pool.active, addr);
-        let is_current_validator = is_current_validator(from);
-        stake_pool.next_epoch_voting_power = stake_pool.next_epoch_voting_power - Coin::value<TestCoin>(&delegation.coins);
-        if (!is_current_validator) {
-            // move to inactive directly if it's not from an active validator
-            IterableTable::add(&mut stake_pool.inactive, &addr, delegation);
+
+        assert!(exists<StakePool>(validator_address), Errors::invalid_argument(ENOT_VALIDATOR));
+        let stake_pool = borrow_global_mut<StakePool>(validator_address);
+
+        assert!(exists<Delegation>(delegator_address), Errors::invalid_argument(EDELEGATION_NOT_FOUND));
+        let delegation = borrow_global_mut<Delegation>(delegator_address);
+        assert!(
+            delegation.validator_address == validator_address,
+            Errors::invalid_argument(EDELEGATION_EXISTS_WITH_DIFFERENT_VALIDATOR),
+        );
+        assert!(
+            delegation.state == DELEGATION_STATE_ACTIVE,
+            Errors::invalid_argument(EDELEGATION_NOT_ACTIVE),
+        );
+
+        let unlocked_coins = extract_from_pool(&mut stake_pool.active, delegation.shares);
+        let unlocked_amount = Coin::value<TestCoin>(&unlocked_coins);
+        stake_pool.next_epoch_voting_power = stake_pool.next_epoch_voting_power - unlocked_amount;
+
+        if (!is_current_validator(validator_address)) {
+            delegation.state = DELEGATION_STATE_INACTIVE;
+            // move to inactive directly if it's not from an active validator.
+            delegation.shares = buy_shares_from_pool(&mut stake_pool.inactive, unlocked_coins);
         } else if (delegation.locked_until_secs < current_time) {
-            // move to pending_inactive if it can be unlocked
-            IterableTable::add(&mut stake_pool.pending_inactive, &addr,  delegation);
+            delegation.state = DELEGATION_STATE_PENDING_INACTIVE;
+            let pending_inactive_pool = &mut stake_pool.pending_inactive;
+            // move to pending_inactive if it can be unlocked.
+            delegation.shares = buy_shares_from_pool(&mut pending_inactive_pool.pool, unlocked_coins);
+            Vector::push_back(&mut pending_inactive_pool.delegator_addresses, delegator_address);
```

Comment:
this doesn't need to be a pool? just being vector<coin> and vector<address>

Reply:
Same reason as pointed out in the previous comment.